File: library/src/main/java/com/github/jokar/multilanguages/library/MultiLanguage.java
package com.github.jokar.multilanguages.library;

import android.content.ComponentCallbacks;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.LocaleList;
import android.util.DisplayMetrics;

import java.util.Locale;

/**
 * Create by JokAr. on 2019-07-08.
 */
public class MultiLanguage {

    private MultiLanguage() {
    }

    public static void initCache(Context context) {
        LanguageCache.init(context);
    }

    public static void init(final Context context) {
        context.registerComponentCallbacks(new ComponentCallbacks() {
            @Override
            public void onConfigurationChanged(Configuration newConfig) {
                MultiLanguage.saveSystemCurrentLanguage(newConfig);
                MultiLanguage.setApplicationLocale(context);
            }

            @Override
            public void onLowMemory() {
            }
        });
    }

    public static void saveSystemCurrentLanguage() {
        LanguageCache.saveSystemCurrentLanguage();
    }

    public static void saveSystemCurrentLanguage(Configuration newConfig) {
        LanguageCache.saveSystemCurrentLanguage(newConfig);
    }

    public static Locale getSystemCurrentLocale() {
        return LanguageCache.getSystemCurrentLocale();
    }

    public static void saveAppLanguage(Context context, MultiLanguage.AppLanguage language) {
        LanguageCache.saveAppLanguage(context, language);
    }

    public static MultiLanguage.AppLanguage getAppLanguage() {
        return LanguageCache.getAppLanguage();
    }

    /**
     * 根据当前配置，生成新的context
     *
     * @param context attach base context
     * @return 配置后的context
     */
    public static Context setLocale(Context context) {
        Locale locale = LanguageCache.getAppLanguageLocale();
        Locale.setDefault(locale);
        Resources resources = context.getResources();
        DisplayMetrics dm = resources.getDisplayMetrics();
        Configuration config = resources.getConfiguration();
        config.locale = locale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                // Android 7.0 后支持设置多个偏好语言
                LocaleList.setDefault(new LocaleList(locale));
            }
            config.setLocale(locale); // Android 4.2 后使用 setLocale 设置语言
            context = context.createConfigurationContext(config);
        } else {
            resources.updateConfiguration(config, dm);
        }
        return context;
    }

    /**
     * 设置application context locale
     *
     * @param context
     */
    public static void setApplicationLocale(Context context) {
        Locale locale = LanguageCache.getAppLanguageLocale();
        Locale.setDefault(locale);
        Resources resources = context.getApplicationContext().getResources();
        DisplayMetrics dm = resources.getDisplayMetrics();
        Configuration config = resources.getConfiguration();
        config.locale = locale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                // Android 7.0 后支持设置多个偏好语言
                LocaleList.setDefault(new LocaleList(locale));
            }
            config.setLocale(locale); // Android 4.2 后使用 setLocale 设置语言
        }
        resources.updateConfiguration(config, dm);
    }

    /**
     * 从configuration中获取系统语言
     *
     * @param newConfig
     * @return
     */
    public static Locale getSystemLocale(Configuration newConfig) {
        Locale locale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            locale = newConfig.getLocales().get(0);
        } else {
            locale = newConfig.locale;
        }
        return locale;
    }

    /**
     * 从locale default中获取系统语言
     *
     * @return
     */
    public static Locale getSystemLocale() {
        Locale locale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            locale = LocaleList.getDefault().get(0);
        } else {
            locale = Locale.getDefault();
        }
        return locale;
    }

    public enum AppLanguage {
        FOLLOW_SYSTEM,
        CHINA,
        TAIWAN,
        ENGLISH
    }
}

File: library/src/main/java/com/github/jokar/multilanguages/library/MultiLanguage.kt
package com.github.jokar.multilanguages.library

import android.content.Context
import android.content.res.Configuration
import android.os.Build
import android.os.LocaleList
import java.util.*

/**
 * Create by JokAr. on 2019-07-08.
 */
object MultiLanguage {
    private var languageLocalListener: LanguageLocalListener? = null
    fun init(listener: LanguageLocalListener?) {
        languageLocalListener = listener
    }

    fun setLocal(context: Context): Context {
        return updateResources(context, getSetLanguageLocale(context))
    }

    /**
     * 设置语言类型
     */
    fun setApplicationLanguage(context: Context) {
        val resources = context.applicationContext.resources
        val dm = resources.displayMetrics
        val config = resources.configuration
        val locale = getSetLanguageLocale(context)
        config.locale = locale
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val localeList = LocaleList(locale)
            LocaleList.setDefault(localeList)
            config.setLocales(localeList)
            context.applicationContext.createConfigurationContext(config)
            Locale.setDefault(locale)
        }
        resources.updateConfiguration(config, dm)
    }

    /**
     * @param context
     */
    fun onConfigurationChanged(context: Context) {
        setLocal(context)
        setApplicationLanguage(context)
    }

    /**
     * 获取选择的语言
     *
     * @param context
     * @return
     */
    private fun getSetLanguageLocale(context: Context): Locale? {
        return if (languageLocalListener != null) {
            languageLocalListener!!.getSetLanguageLocale(context)
        } else Locale.ENGLISH
    }

    /**
     * 更新语言设置
     *
     * @param context
     * @param locale
     * @return
     */
    private fun updateResources(context: Context, locale: Locale?): Context {
        var context = context
        Locale.setDefault(locale)
        val res = context.resources
        val config = Configuration(res.configuration)
        if (Build.VERSION.SDK_INT >= 17) {
            config.setLocale(locale)
            context = context.createConfigurationContext(config)
        } else {
            config.locale = locale
            res.updateConfiguration(config, res.displayMetrics)
        }
        return context
    }

    /**
     * 获取系统语言
     * @param newConfig
     * @return
     */
    fun getSystemLocal(newConfig: Configuration): Locale {
        val locale: Locale
        locale = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            newConfig.locales[0]
        } else {
            newConfig.locale
        }
        return locale
    }

    /**
     * 获取系统语言
     * @param context
     * @return
     */
    fun getSystemLocal(context: Context?): Locale {
        val locale: Locale
        locale = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            LocaleList.getDefault()[0]
        } else {
            Locale.getDefault()
        }
        return locale
    }
}